### PR TITLE
Early version of fx graph matcher for NS

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52130 [wip] ns for fx: add support for subgraph matching
* #52092 NS for FX: add test for a simple sparsenn model
* #51669 ns for fx - stubs of the three APIs (compare weights, activations, activations with shadow)
* **#51588 Early version of fx graph matcher for NS**

Summary:

Early version of utility to match nodes between graph A and graph B, for Numerical Suite for FX graph mode quantization.

The main goal of this utility is to reliably match the nodes of graph A to the nodes of graph B, and throw an easy to read error message.  This will be used in future PRs to create the APIs for matching activations.  It also could potentially be used to match weights.

Test Plan:

For now, we have bare bones test coverage on some toy models, and a single torchvision model.

```
python test/test_quantization.py TestFXGraphMatcher
```

Future PRs will add more testing.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403093](https://our.internmc.facebook.com/intern/diff/D26403093)

### CODE REVIEW

File: torch/quantization/ns/graph_matcher.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        Returns the next matchable node.
+        """
+        while len(self.stack) > 0:
+            cur_node = self.stack.pop()
+            if cur_node in self.seen_nodes:
+                continue
+            self.seen_nodes.add(cur_node)
+            # add args of previous nodes to stack
+            # TODO(future PR): handle kwargs as needed
+            for arg in cur_node.args:
```
Comment (raghuramank100) [inline torch/quantization/ns/graph_matcher.py:137]: For non matchable nodes, we may need to do special handling when we append inputs to the stack.

File: test/quantization/test_numeric_suite_fx.py [hunk 2]
```diff
@@ -429,3 +443,117 @@ def test_compare_model_outputs_lstm_dynamic_fx(self):
...
+        mp = prepare_fx(m, {'': torch.quantization.default_qconfig})
+        # TODO(future PR): prevent the need for copying here, we can copy the
+        # modules but should reuse the underlying tensors
+        mp_copy = copy.deepcopy(mp)
+        mq = convert_fx(mp_copy)
+        # assume success if no exceptions
+        results = get_matching_node_pairs(mp, mq)
+
+    @skip_if_no_torchvision
+    def test_mobilenet_v2_fp32_qat_prepared_vs_int8(self):
```
Comment (raghuramank100) [inline test/quantization/test_numeric_suite_fx.py:549]: For later PRs, can we also check matching for fp32 vs prepared and prepared QAT?

File: torch/quantization/ns/graph_matcher.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+        ]),
+        # add
+        set([
+            torch.add,
+            toq.add,
+            operator.add,  # x + y
+        ]),
+    ]
+
+    type_a_related_to_b: Set[Tuple[Callable, Callable]] = set()
```
Comment (raghuramank100) [inline torch/quantization/ns/graph_matcher.py:61]: Noob question: Cant we just verify that for 2 modules, both belong to the same set. Do we need a set of tuples here?
i.e for two graphs to match, each node i in the graph should belong to the same set in the sets_of_related_ops

File: torch/quantization/ns/graph_matcher.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    if node_a.op != node_a.op:
+        # for now, comparing call_module to call_function is not supported
+        # this can be added later if needed
+        return False
+
+    if node_a.op == 'call_function':
+        if node_a.target == node_b.target:
+            # nodes with equivalent targets always match (i.e. F.linear and F.linear)
+            return True
+        key = (node_a.target, node_b.target)
```
Comment (raghuramank100) [inline torch/quantization/ns/graph_matcher.py:185]: Alternately if both node_a.target and node_b.target belong to the same set, it could work.

File: test/quantization/test_numeric_suite_fx.py (hunk 2 above)
Comment (vkuzo) [reply test/quantization/test_numeric_suite_fx.py:549]: yes, definitely

File: torch/quantization/ns/graph_matcher.py (hunk 3 above)
Comment (vkuzo) [reply torch/quantization/ns/graph_matcher.py:61]: The current code precomputes "x and y are in the same set" instead of doing that at inference time.  I'm flexible though, this could be done in multiple correct ways, I don't feel strongly.

File: torch/quantization/ns/graph_matcher.py (hunk 1 above)
Comment (vkuzo) [reply torch/quantization/ns/graph_matcher.py:137]: hmm, would you mind providing more context?  Not sure what you mean.

Let's say a is matchable and b is non_matchable.  If we have a graph of

```
a0 -> b0 -> a1 -> b1
```

And we iterate from b1 backwards, the current code will add a1 to the stack when we are in b1, and add a0 to the stack when we are in b0, etc.

File: torch/quantization/ns/graph_matcher.py (hunk 4 above)
Comment (vkuzo) [reply torch/quantization/ns/graph_matcher.py:185]: sure, I don't feel strongly about how we implement this as long as it works

File: torch/quantization/ns/graph_matcher.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    pass
+
+def _node_a_related_to_b(
+    node_a: Node,
+    node_b: Node,
+    gm_a: GraphModule,
+    gm_b: GraphModule,
+    type_a_related_to_b: Set[Tuple[Callable, Callable]],
+) -> bool:
+    if node_a.op != node_a.op:
```
Comment (hx89) [inline torch/quantization/ns/graph_matcher.py:176]: Do you mean:
    if node_a.op != node_b.op:

File: test/quantization/test_numeric_suite_fx.py [hunk 6]
```diff
@@ -429,3 +443,117 @@ def test_compare_model_outputs_lstm_dynamic_fx(self):
             lstm_input,
             lstm_hidden,
         )
+
+class TestFXGraphMatcher(QuantizationTestCase):
+    # TODO(future PR): more tests
+
+    def test_conv_mod_fp32_prepared_vs_int8(self):
```
Comment (hx89) [inline test/quantization/test_numeric_suite_fx.py:450]: Can we add test case for matching fused ops like float and quantized ConvRelu in future PRs?

Comment (vkuzo) [reply test/quantization/test_numeric_suite_fx.py:450]: sure, will do

File: torch/quantization/ns/graph_matcher.py (hunk 5 above)
Comment (vkuzo) [reply torch/quantization/ns/graph_matcher.py:176]: yes, thanks, good catch :)

File: test/quantization/test_numeric_suite_fx.py [hunk 7]
```diff
@@ -429,3 +442,117 @@ def test_compare_model_outputs_lstm_dynamic_fx(self):
...
+        # TODO(future PR): prevent the need for copying here, we can copy the
+        # modules but should reuse the underlying tensors
+        mp_copy = copy.deepcopy(mp)
+        mq = convert_fx(mp_copy)
+        results = get_matching_node_pairs(mp, mq)
+
+        expected_types = {'linear_1': (F.linear, toq.linear)}
+        self.assert_types_for_matched_node_pairs(results, expected_types, mp, mq)
+
+    def test_matching_failure_node_count(self):
```
Comment (jerryzh168) [inline test/quantization/test_numeric_suite_fx.py:482]: nit: can we spilt the test cases to:
- TestFXGraphMatcher: general functionality test
- TestFXGraphMatcherOps: test for individual or fused ops
- TestFXGraphMatcherModels: test for small or large models
Also I saw "fp32_prepared_vs_int8" appear multiple times, I think this is related to TestFXGraphMatcherOps/TestFXGraphMatcherModels maybe we can have separate test case for each pair of comparison as well

File: test/quantization/test_numeric_suite_fx.py [hunk 8]
```diff
@@ -429,3 +442,117 @@ def test_compare_model_outputs_lstm_dynamic_fx(self):
...
+                return z
+
+        m = M().eval()
+        mp = prepare_fx(m, {'': torch.quantization.default_qconfig})
+        # TODO(future PR): prevent the need for copying here, we can copy the
+        # modules but should reuse the underlying tensors
+        mp_copy = copy.deepcopy(mp)
+        mq = convert_fx(mp_copy)
+        # assume success if no exceptions
+        results = get_matching_node_pairs(mp, mq)
```
Comment (jerryzh168) [inline test/quantization/test_numeric_suite_fx.py:532]: I think we might want to verify the result as well, but could be in a separate PR

File: torch/quantization/ns/graph_matcher.py [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+def _get_node_target_type(node: Node, gm: GraphModule) -> Optional[Callable]:
+    if node.op == 'call_function':
+        return node.target  # type: ignore
+    elif node.op == 'call_module':
+        assert isinstance(node.target, str)
+        mod = _getattr_from_fqn(gm, node.target)
+        return type(mod)
+    return None
+
+def get_matching_node_pairs(
```
Comment (jerryzh168) [inline torch/quantization/ns/graph_matcher.py:220]: qq: have we explored whether we can use https://github.com/pytorch/pytorch/blob/master/torch/fx/subgraph_rewriter.py ?

Comment (vkuzo) [reply torch/quantization/ns/graph_matcher.py:220]: yes, let's ask the FX folks

File: test/quantization/test_numeric_suite_fx.py (hunk 8 above)
Comment (vkuzo) [reply test/quantization/test_numeric_suite_fx.py:532]: sg, that might be good to push to a separate PR

File: test/quantization/test_numeric_suite_fx.py (hunk 7 above)
Comment (vkuzo) [reply test/quantization/test_numeric_suite_fx.py:482]: sure, will do in this PR

File: torch/quantization/ns/graph_matcher.py (hunk 9 above)
Comment (vkuzo) [reply torch/quantization/ns/graph_matcher.py:220]: looks like it might take a few days to align on this, happy to update it in a future PR if there is a good fit.

File: test/quantization/test_numeric_suite_fx.py (hunk 7 above)
Comment (vkuzo) [reply test/quantization/test_numeric_suite_fx.py:482]: > Also I saw "fp32_prepared_vs_int8" appear multiple times, I think this is related to TestFXGraphMatcherOps/TestFXGraphMatcherModels maybe we can have separate test case for each pair of comparison as well

oh, by the way, "fp32_prepared" vs "int8" is the only pair of models which is expected to work with the current code.  The code should be generic to allow int8 vs fp32_prepared (switched order), as well as adding more model types.  That is not tested yet and saved for future PRs.

**Review by jerryzh168 — APPROVED** [review]
I feel the current code can't use subgraph_rewriter, but if we extend this to support subgraph matching then we might need that.
